Add MeasurementTimescales Tag

wthrowe opened the pull request:
Currently the values are set to a fairly arbitrary unexpiring
constant.  Adjusting them will need feedback from parts of the control
system that are not yet complete.

## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe committed:
Add MeasurementTimescales Tag

Currently the values are set to a fairly arbitrary unexpiring
constant.  Adjusting them will need feedback from parts of the control
system that are not yet complete.

knelli2 commented on src/ControlSystem/Tags.hpp:
```diff
@@ -0,0 +1,56 @@
+    for (const auto& function_of_time : domain_creator->functions_of_time()) {
+      const double function_initial_time =
+          function_of_time.second->time_bounds()[0];
+      auto initial_time_scale = make_with_value<DataVector>(
+          function_of_time.second->func(function_initial_time)[0], 0.1);
+      timescales.emplace(
+          function_of_time.first,
```
(optional) Even though this map of unique pointers is closely related to the functions of time from the domain creator, for diagnostic and clarity purposes they should probably have different (but related) names. Maybe prepending or appending "MeasurementTimescale" to `function_of_time.first`?

wthrowe replied:
That'd make it kind of unpleasant if we ever need to look up the normal FoTs when looping over the timescales.

knelli2 replied:
Hmm yes true. Ok leave as is then.

knelli2 reviewed (commented):
Just an optional comment about names. If you choose to do it you can squash right away, otherwise I approve.

nilsvu commented on src/ControlSystem/Tags.hpp:
```diff
@@ -0,0 +1,56 @@
+    std::unordered_map<std::string,
+                       std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>
+        timescales;
+    for (const auto& function_of_time : domain_creator->functions_of_time()) {
```
use structured bindings: `const auto& [f_of_t_name, function_of_time]`?

wthrowe replied:
I keep finding more reasons to dislike structured bindings.  Most recently: can't read them in gdb.  If you insist I'll change it, but otherwise I'm leaving this as-is.

nilsvu replied:
Nope, up to you

nilsvu commented on src/ControlSystem/Tags.hpp:
```diff
@@ -0,0 +1,56 @@
+      const double function_initial_time =
+          function_of_time.second->time_bounds()[0];
+      auto initial_time_scale = make_with_value<DataVector>(
+          function_of_time.second->func(function_initial_time)[0], 0.1);
```
Why `0.1` here?

Also, I suggest you write `const DataVector used_for_size = function_of_time->func(function_initial_time)[0];` and pass it to `make_with_value`, just to make clear the values don't matter.

wthrowe replied:
0.1 was arbitrary since the actual control systems don't exist yet to set it.  Actually, though, setting it to the initial step size is less arbitrary and might even make sense as a real initial value.

nilsvu commented on src/ControlSystem/Tags.hpp:
```diff
@@ -0,0 +1,56 @@
+
+namespace control_system::Tags {
+/// The measurement timescales associated with
+/// domain::Tags::FunctionsOfTime.
```
Could you give some more details here, like how you represent the timescales as `PiecewisePolynomial<0>`?

wthrowe commented:
In addition to addressing the comments, I have adjusted the code so that it does not create measurement timescale entries for functions of time that never expire, since those cannot be modified by a control system and so shouldn't be measured.

nilsvu reviewed (dismissed):
LGTM, feel free to squash

knelli2 commented on src/ControlSystem/Tags.hpp:
```diff
@@ -0,0 +1,80 @@
+
+      // This check is intentionally inside the loop over the
+      // functions of time so that it will not trigger for domains
+      // without control systems.
+      if (initial_time_step <= 0.0) {
+        ERROR(
```
LCOV

wthrowe replied:
Added tests.  New error test and the last block in the preexisting test.

knelli2 commented on tests/Unit/ControlSystem/Test_Tags.cpp:
```diff
@@ -0,0 +1,82 @@
+#include "Utilities/ErrorHandling/Error.hpp"
+#include "Utilities/Literals.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace {
+static const double initial_time = 2.0;
```
Clang-tidy complains about the `static` being redundant.

knelli2 reviewed (commented):
Looks good to me as well. Squash these in directly.

wthrowe reviewed (commented):
Squashed, including new things.  Rebased on develop.
